File: src/main/java/org/gjt/jclasslib/structures/attributes/TypePathEntry.java
/*
 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public
 License as published by the Free Software Foundation; either
 version 2 of the license, or (at your option) any later version.
 */

package org.gjt.jclasslib.structures.attributes;

import org.gjt.jclasslib.structures.AbstractStructure;
import org.gjt.jclasslib.structures.InvalidByteCodeException;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;

public class TypePathEntry extends AbstractStructure {

    private int typePathKind;
    private int typeArgumentIndex;

    public int getTypePathKind() {
        return typePathKind;
    }

    public void setTypePathKind(int typePathKind) {
        this.typePathKind = typePathKind;
    }

    public int getTypeArgumentIndex() {
        return typeArgumentIndex;
    }

    public void setTypeArgumentIndex(int typeArgumentIndex) {
        this.typeArgumentIndex = typeArgumentIndex;
    }

    @Override
    public void read(DataInput in) throws InvalidByteCodeException, IOException {
        super.read(in);
        typePathKind = in.readUnsignedByte();
        typeArgumentIndex = in.readUnsignedByte();
    }

    @Override
    public void write(DataOutput out) throws InvalidByteCodeException, IOException {
        super.write(out);
        out.writeByte(typePathKind);
        out.writeByte(typeArgumentIndex);
    }

    @Override
    protected String printAccessFlagsVerbose(int accessFlags) {
        return "";
    }
}

File: modules/data/src/main/kotlin/structures/attributes/TypePathEntry.kt
/*
 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public
 License as published by the Free Software Foundation; either
 version 2 of the license, or (at your option) any later version.
 */

package org.gjt.jclasslib.structures.attributes

import java.io.DataInput
import java.io.DataOutput

/**
 * Type path entry for a TypeAnnotation structure.
 */
class TypePathEntry private constructor() : SubStructure() {

    /**
     * Constructor.
     * @param typePathKind the type path kind
     */
    constructor(typePathKind: TypePathKind) : this() {
        this.typePathKind = typePathKind
    }

    internal constructor(input: DataInput) : this() {
        read(input)
    }

    /**
     * The type path kind.
     */
    lateinit var typePathKind: TypePathKind

    /**
     * The index of the type argument.
     */
    var typeArgumentIndex: Int = 0

    override fun readData(input: DataInput) {
        typePathKind = TypePathKind.getFromTag(input.readUnsignedByte())
        typeArgumentIndex = input.readUnsignedByte()
    }

    override fun writeData(output: DataOutput) {
        output.writeByte(typePathKind.tag)
        output.writeByte(typeArgumentIndex)
    }

    override val debugInfo: String
        get() = "with typePathKind $typePathKind, typeArgumentIndex $typeArgumentIndex"

    override val length: Int
        get() = 2
}
